Problem: We could not encode non-Foundation objects

hamchapman opened the pull request:
Solution: Add a `CBORRepresentable` protocol, which has a single required function named `getCBORRepresentation`. This must return an `NSObject *`. This returned object is what is then used to try and encode the object into CBOR.

---

~~Gonna add a little test for this but asking for review now because it's just a small change and the test isn't the bit I want input/feedback 😁~~

hamchapman commented:
I've updated it to use the protocol approach. I like it. Part of the reason I did the quick and dirty version before was because I'd overlooked that the repo was public. Definitely makes more sense to have something more extensible if we're keeping this open 👍

hamchapman commented:
Not sure what you think the best way to handle the header situation is. I've currently got `CBORRepresentable.h` as a separate header that's also getting copied into the build's `include` directory and so needs to be imported separately. Maybe it makes more sense to have `ObjCCBOR.h` import it so that there's a single header import required?

connorpower commented:
> Not sure what you think the best way to handle the header situation is.

To be totally future proof and standard, we should probably convert `ObjCCBOR.h` into an umbrella header and move whatever is currently in there into a new header file (plus rename `.m` file). Given the size of the library, I wouldn't be opposed to just throwing everything directly into the `ObjCCBOR.h` file either though.

hamchapman commented:
Okay I've updated this to make `ObjCCBOR.h` an umbrella header, as suggested.

That meant, as you already knew, a renaming of the existing `ObjCCBOR.{h,m}`. 

I went with just `CBOR.{h,m}` and changed the class name while I was at it, so the calls are now `[CBOR encode]` and `[CBOR decode]` etc

Let me know what you think about it 👍

connorpower commented on ObjCCBOR/CBORRepresentable.h:
```diff
@@ -0,0 +1,19 @@
+
+@protocol CBORRepresentable
+
+- (nullable NSObject *)getCBORRepresentation;
```
```suggestion
- (nullable NSObject *)CBORRepresentation;
```

This was bugging me for a while and (although I suggested it) I've finally worked out why. Working in so many languages at once can really scramble your brain 🥴

> If the method returns an attribute of the receiver, name the method after the attribute. The use of “get” is unnecessary, unless one or more values are returned indirectly.

[Coding Guidelines for Cocoa](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html)

connorpower commented on ObjCCBOR/CBOR.h:
```diff
@@ -0,0 +1,44 @@
+ `NSNumber`, `NSNull` or `NSData`. Will return `nil` if the decoding
+ fails.
+ */
++ (nullable id)decode:(NSData *)data error:(NSError * _Nullable __autoreleasing *)error;
```
```suggestion
+ (nullable id)decodeData:(NSData *)data error:(NSError * _Nullable __autoreleasing *)error;
```

> Make the word before the argument describe the argument.
> 
> |Example||
> |-- | -- |
> | - (id)viewWithTag:(NSInteger)aTag; | Right. |
> | - (id)taggedView:(int)aTag; | Wrong. |

[Coding Guidelines for Cocoa](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html)

---

When I looked a little closer, I realized `NSJSONSerialization` does this too.

connorpower commented on ObjCCBOR/CBOR.h:
```diff
@@ -0,0 +1,44 @@
+ an error object will be written to this parameter should encoding fail.
+ @returns Returns the encoded data or `nil` if the encoding fails.
+ */
++ (NSData *)encode:(NSObject *)object error:(NSError * _Nullable __autoreleasing *)error;
```
```suggestion
+ (NSData *)encodeObject:(NSObject *)object error:(NSError * _Nullable __autoreleasing *)error;
```

> Make the word before the argument describe the argument.
> 
> |Example||
> |-- | -- |
> | - (id)viewWithTag:(NSInteger)aTag; | Right. |
> | - (id)taggedView:(int)aTag; | Wrong. |

[Coding Guidelines for Cocoa](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html)

---

When I looked a little closer, I realized `NSJSONSerialization` does this too.

connorpower commented on ObjCCBOR/ObjCCBOR.h:
```diff
@@ -2,43 +2,11 @@
-@end
-
-NS_ASSUME_NONNULL_END
+#import "CBOR.h"
+#import "CBORRepresentable.h"
```
In the umbrella header, these should use angle bracket syntax.

```suggestion
#import <ObjCCBOR/CBOR.h>
#import <ObjCCBOR/CBORRepresentable.h>
```

---

As an aside: choosing the framework header over simply including everything in there was definitely the right move.

> Including framework header files. Each framework should have a header file, named after the framework, that includes all the public header files of the framework.
> 
> Header file | Framework
> -- | --
> Foundation.h | Foundation.framework.
> 

[Coding Guidelines for Cocoa](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingBasics.html#//apple_ref/doc/uid/20001281-1002806-BBCFIFEH)
